src/impl/collection_change_builder.hpp

```diff
@@ -39,24 +41,38 @@ class CollectionChangeBuilder : public CollectionChangeSet {
 
     // Calculate where rows need to be inserted or deleted from old_rows to turn
     // it into new_rows, and check all matching rows for modifications
+    // If `move_candidates` is supplied they it will eb used to do more accurate
+    // determination of which rows moved. This is only supported when the rows
+    // are in table order (i.e. not sorted or from a LinkList)
     static CollectionChangeBuilder calculate(std::vector<size_t> const& old_rows,
                                              std::vector<size_t> const& new_rows,
                                              std::function<bool (size_t)> row_did_change,
-                                             bool sort);
+                                             util::Optional<IndexSet> const& move_candidates = util::none);
 
+    // generic operations {
+    CollectionChangeSet finalize() &&;
     void merge(CollectionChangeBuilder&&);
-    void clean_up_stale_moves();
 
     void insert(size_t ndx, size_t count=1, bool track_moves=true);
     void modify(size_t ndx);
     void erase(size_t ndx);
-    void move_over(size_t ndx, size_t last_ndx, bool track_moves=true);
     void clear(size_t old_size);
+    // }
+
+    // operations only implemented for LinkList semantics {
+    void clean_up_stale_moves();
     void move(size_t from, size_t to);
+    // }
 
-    void parse_complete();
+    // operations only implemented for Row semantics {
+    void move_over(size_t ndx, size_t last_ndx, bool track_moves=true);
+    // must be followed by move_over(old_ndx, ...)
+    // precondition: `new_ndx` must be a new insertion
+    void subsume(size_t old_ndx, size_t new_ndx, bool track_moves=true);
+    void swap(size_t ndx_1, size_t ndx_2, bool track_moves=true);
 
-    CollectionChangeSet finalize() &&;
+    void parse_complete();
```

Comment:
Sort of? It currently only needs to be called for row parsing since it transforms the move tracking for that into the move tracking for LinkLists (and the output).